### PR TITLE
Custom user plugin [fix #574]

### LINKED ISSUE #574: [Plugin] Move plugin definition to external Json file
Hello, 

Recently, I was thinking about the possibility for the users to have additional plugin that are not officially integrated to brainstorm (typically code that are in .brainstorm/process). For example, I have a lot of code related to nirs project that I sometime needs to share with colleague and currently it makes it hard to track version of the scripts. With a plugin system, I could make sure all the script are always up-to-date between me and my colleagues and between my different computers :)  

For the main plugin, this would allow some user to change the version of the external-software they want to use more easily.

To do that I was thinking to move the definition of the plugin to two files: but_plugin.json in the main brainstorm folder. and allows users to create a user_plugin.json in .brainstorm folder to specify user-specific 

https://github.com/brainstorm-tools/brainstorm3/blob/master/toolbox/core/bst_plugin.m#L128-L599 would then be only reading of the json files. 

Here is a proof of concept of how the file would looks like
```
PlugDesc            = struct();
PlugDesc.Name           = 'bci2000';
PlugDesc.Version        = 'latest';
PlugDesc.Category       = 'I/O';
PlugDesc.URLzip         = 'https://bci2000.org/downloads/mex.zip';
PlugDesc.URLinfo        = 'https://www.bci2000.org/mediawiki/index.php/User_Reference:Matlab_MEX_Files';
PlugDesc.TestFile       = 'load_bcidat.m';
PlugDesc.CompiledStatus = 0;

txt = jsonencode(PlugDesc,PrettyPrint=true)


value = jsondecode(txt);
PlugDescDecode = struct_copy_fields(bst_plugin('GetStruct','bci2000'), value, 1); % same as existing struct
```

Let me know if you think that is a good idea. I can start working on a PR then. 

Edouard

### PR DESCRIPTION
**Pull request by Edouard2laire**
Hello, 

As discussed here: https://github.com/brainstorm-tools/brainstorm3/issues/574 , this allows the usage of custom plugins that are not officially supported in Brainstorm.  This can be used to share processes within a lab by having a common GitHub repository and having that repo as a custom plugin in Brainstorm.

This is done by defining the customs plugin in specific JSON files under the .brainstorm folder. Each json can contain one or multiple plugins (although one plugin per json might be better as recommended by Francois)  

Here is an example of plugins.json file with 2 plugin definition: 
[plugins.json](https://github.com/brainstorm-tools/brainstorm3/files/14289268/plugins.json)

Splitted into two files: 

[plugin_bids_matlab.json](https://github.com/brainstorm-tools/brainstorm3/files/14289410/plugin_bids_matlab.json)
[plugin_bst_users.json](https://github.com/brainstorm-tools/brainstorm3/files/14289411/plugin_bst_users.json)

The creation of the JSON file can be a bit tedious and prone to error but this is just the first PR to get feedback and we might improve the addition of a plugin in that json file later :)   As you suggested, made json file could easily be shared in the bst_plugin repo. 

Edouard

[Hs: i cant believe i posted that issue in 2022] 

TODOs
- [x] Update plugin menu without restarting Brainstorm
- [x] Add user-defined plugins with URL
- [x] Add user-defined plugins with file
- [x] Add user-defined plugins manually (user input)
- [x] Remove user-defined plugins
- [x] Tests with archiving
- [x] Tests with deploy
- [x] Documentation

### CODE REVIEW

**Comment by rcassani** [comment]
Can we save the user from creating the `json` file?

If the scripts to be shared are a public repo, and they are thought to be used as Brainstorm processes. This can work as:

1. An option in the GUI. Plugin > Add from URL
2. User provides the URL to the repo (may branch as well)
3. The repo is downloaded
4. Brainstorm checks for the `json` file in it
5. If the `json` file exists the plugin gets added following those instructions

**Comment by Edouard2laire** [comment]
> Can we save the user from creating the `json` file?

yes, I just wanted to make sure you were still ok with the idea since it is from 2022 before working on the user-friendly part :) 


> If the scripts to be shared are a public repo, and they are thought to be used as Brainstorm processes. This can work as:
> 
> 1. An option in the GUI. Plugin > Add from URL
> 2. User provides the URL to the repo (may branch as well)
> 3. The repo is downloaded
> 4. Brainstorm checks for the `json` file in it
> 5. If the `json` file exists the plugin gets added following those instructions

Seems like a good plan, and if the json file doesn't exist then we ask the user to fill it being extra cautious.  I would also add the option to Add directly from json in addition to the URL from step 1.

**Comment by rcassani** [comment]
We can retrieve only the `.json` file, something like this:

```
text = bst_webread('https://raw.githubusercontent.com/brainstorm-tools/brainstorm3/master/README.md');
```

**Comment by Edouard2laire** [comment]
Here is a first implementation. The following code is adding the plugin in the .brainstorm folder. 

bst_plugin('AddCustom','https://github.com/Edouard2laire/bst-users/blob/custom_plugin/plugins/plugin_bids_matlab.json')


Is there a way to update the plugin menu without starting brainstorm again ?

**Comment by Edouard2laire** [comment]
ok . This is ready to be reviewed. The only issue is that when adding a new plugin, brainstorm needs to be restarted to update the menu.

**Comment by Edouard2laire** [comment]
is there any news for that PR by any chance ? :)

**Comment by rcassani** [comment]
Just started with it

File: toolbox/core/bst_plugin.m [hunk 1]
```diff
@@ -645,16 +645,14 @@
...
-            continue;
-        end
-
-        PlugDesc  = [PlugDesc , PlugDesc_tmp];
+    % === USER DEFINED PLUGINS ===
+    plugJsonFiles = dir(fullfile(bst_get('BrainstormUserDir'), 'plugin_*.json'));
+    for ix = 1:length(plugJsonFiles)
+        plugJsonText = fileread(fullfile(plugJsonFiles(ix).folder, plugJsonFiles(ix).name));
+        PlugDesc_tmp = bst_jsondecode(plugJsonText);
+        PlugDesc(end+1) = GetStruct(PlugDesc_tmp.Name);
```
Comment (Edouard2laire) [inline toolbox/core/bst_plugin.m:653]: Would ``` PlugDesc(end+1) = GetStruct(lower(PlugDesc_tmp.Name)); be better here ? I think I had issue when testing with plugin having capital letter in their name

Comment (rcassani) [reply toolbox/core/bst_plugin.m:653]: At this point the name should be sanitized in the json file by the validation see: [8317475](https://github.com/brainstorm-tools/brainstorm3/pull/683/commits/8317475cbe43b4ffd430cdc580c3bd665d9473e2)

**Comment by rcassani** [comment]
@Edouard2laire, I'm still working on this PR. Meanwhile, what do you think on deleting the json file (in the plugin dir) when uninstall a 'User defined' plugin? I think it makes sense, as if they are not needed for the moment, they can be unloaded. Otherwise, a method to delete the json files would be needed.

**Comment by Edouard2laire** [comment]
> @Edouard2laire, I'm still working on this PR. Meanwhile, what do you think on deleting the json file (in the plugin dir) when uninstall a 'User defined' plugin? I think it makes sense, as if they are not needed for the moment, they can be unloaded. Otherwise, a method to delete the json files would be needed.

yes, I think that make sense too. Users can easily add back the plugin if they want.

**Comment by rcassani** [comment]
> yes, I think that make sense too. Users can easily add back the plugin if they want.

It looked as a good idea, but now it is quite asymmetric. 
"Add", just adds, Install is needed
'Uninstall', uninstalls and deletes the json file

I'll add the option of remove the custom plugin:
* Add > Install
* Uninstall > Remove (with remove uninstalling if needed)

**Comment by Edouard2laire** [comment]
Wow it looks great! thanks a lot. (sorry I didn't realize there was so much code left to write)

For the addition of new plugin, do we need to have 3 options ( file, URL, and manual)? I have the feeling that the file and URL would already cover most of the options.

**Comment by rcassani** [comment]
Wanted to add the manual, as a way to start create a basic .json without having to type it.

In addition, documentation is needed for this new feature. I'll start with draft and share it with you.
Do you have access to edit pages in the Brainstorm wiki?

**Comment by Edouard2laire** [comment]
> Wanted to add the manual, as a way to start create a basic .json without having to type it.
> 
> In addition, documentation is needed for this new feature. I'll start with draft and share it with you. Do you have access to edit pages in the Brainstorm wiki?

Yes, I have access (username: EdouardDelaire).  Let me know if there is anything else you want me to do.

**Comment by rcassani** [comment]
@Edouard2laire, documentation is in the link below, please check it out and edit it if needed.
https://neuroimage.usc.edu/brainstorm/Tutorials/Plugins

This PR seems to be read to be merged!

**Comment by Edouard2laire** [comment]
Looks good to me. Thx :)

**Comment by rcassani** [comment]
@Edouard2laire 🎉 
(and before 2026 🤣)

**Comment by Edouard2laire** [comment]
🎉🎉
Thanks for your help in this PR :) 
(hopefully, I'll have graduated by 2026 lol)

**Pull request merged**